<launch>
    <param name="use_sim_time" value="false"/>

    <arg name="gazebo" default="true"/>
    <arg name="vision" default="true"/>
    <arg name="optimizer" default="true"/>
    <arg name="logger" default="false"/>
    <arg name="vision" default="true"/>
    <arg name="is_model_floating_base" value="true"/>
    <arg name="opt_interpolation_time" value="0.05"/>
    <arg name="ctrl_interpolation_time" value="0.01"/>

    <!-- xbotcore -->
    <node if="$(arg gazebo)" pkg="teb_test" type="xbotcore.sh" name="xbot" output="screen"/>

    <group if="$(arg vision)">
        <!-- vision node -->
        <node pkg="teb_test" type="vision_node" name="vision_node" output="screen" args="/D435i/D435i_camera/depth/color/points"/>
<!--        <node pkg="teb_test" type="vision_node" name="vision_node" output="screen" args="/velodyne_points"/>-->

        <node pkg="octomap_server" type="octomap_server_node" name="octomap_server">
                <param name="resolution" value="0.05" />
                <param name="frame_id" type="string" value="world" />
                <param name="sensor_model/max_range" value="1.5" />
<!--                <param name="filter_ground" value="true"/>-->
<!--                <param name="base_frame_id" value="world"/>-->
<!--                <param name="ground_filter/distance" value = "0.1"/>-->
<!--                <param name="ground_filter/plane_distance" value="-0.65"/>-->
                <param name="latch" value="false"/>
                <remap from="cloud_in" to="/planar" />
        </node>

        <!-- robot_body_filter node -->
<!--        <node pkg="robot_body_filter_node" type="robot_body_filter_node_node" name="robot_body_filter_node">
          <rosparam command="load" file="$(find robot_body_filter_node)/config/config.yaml" />
          <remap from="cloud_in" to="/voxel" />
        </node>-->
    </group>

<!--    <group if="$(arg optimizer)">-->
        <param name="cartesian_stack" textfile="$(find teb_test)/config/problem/centauro_stack.yaml"/>
        <param name="collision_urdf" textfile="$(find centauro_urdf)/urdf/centauro_capsule.urdf"/>
        <param name="robot_description_semantic" textfile="$(find centauro_srdf)/srdf/centauro.srdf"/>

        <!-- include robot_description -->
        <include file="$(find centauro_gazebo)/launch/centauro_world.launch">
            <arg name="floating_base" value="true"/>
            <arg name="upper_body" value="true"/>
            <arg name="legs" value="true"/>
            <arg name="realsense" value="true"/>
            <arg name="velodyne" value="true"/>
            <arg name="gazebo" value="$(arg gazebo)"/>
        </include>

        <!-- URDF joint fixer -->
        <node pkg="teb_test" type="urdf_parser_test_node" name="urdf_parser_test_node" output="screen">
            <rosparam param="fixed_joints" command="load" file="$(find teb_test)/config/problem/fixed_joints_car_model.yaml"/>
        </node>

        <!-- optimizer node -->
        <node pkg="teb_test" type="car_model" name="optimizer" output="screen">
            <param name="rate" value="10"/>
            <param name="optimizer_config" textfile="$(find teb_test)/config/problem/constraint.yaml"/>
            <param name="is_model_floating_base" value="$(arg is_model_floating_base)"/>
            <param name="planner_config" textfile="$(find teb_test)/config/problem/planner_config.yaml"/>
            <param name="interpolation_time" type="double" value="0.05"/>
            <rosparam param="fixed_joints" command="load" file="$(find teb_test)/config/problem/fixed_joints.yaml"/>
        </node>

        <!-- robot control node -->
        <node pkg="teb_test" type="robot_controller_car_model_node" name="robot_controller_node" output="screen" launch-prefix="bash -c 'sleep 7.0; $0 $@' ">
            <param name="is_model_floating_base" value="$(arg is_model_floating_base)"/>
            <param name="model_type" value="RBDL"/>
            <param name="opt_interpolation_time" type="double" value="0.1"/>
            <param name="ctrl_interpolation_time" type="double" value="0.01"/>
            <rosparam param="fixed_joints" command="load" file="$(find teb_test)/config/problem/fixed_joints.yaml"/>
        </node>

        <!-- rviz -->
        <node pkg="rviz" name="rviz" type="rviz" output="screen"
            args="-d $(find teb_test)/config/rviz/car_model.rviz"/>

<!--    </group>-->

    <!-- logger node -->
<!--    <node if="$(arg logger)" pkg="multi_contact_planning" type="logger_node" name="logger_node" output="screen"-->
<!--        args="/optimizer/error-->
<!--              /optimizer/time-->
<!--              /xbotcore/joint_states"/>-->
</launch>
